### PR TITLE
Allows embedded keys with data

### PR DESCRIPTION
**Pull request by alecklandgraf**
Add support for using RTA with duplicate values. 

Say you want `beer`, `happy_hour`, and `draft` to return the same emoji, 🍺, this change allows that.

I also fixed what I think was a invalid check of `data`'s type.

### CODE REVIEW

File: src/Textarea.jsx [hunk 1]
```diff
@@ -261,7 +259,7 @@ class ReactTextareaAutocomplete extends React.Component<
 
     providedData
       .then(data => {
-        if (typeof providedData !== 'object') {
```
Comment (alecklandgraf) [inline src/Textarea.jsx]: I think `data` is what should be checked to be an array, `providedData` will always be a object if it's a Promise

Comment (jukben) [reply src/Textarea.jsx]: Yeah! You are definitely right. I'm just wondering why not use `Array.isArray`?

File: src/List.jsx [hunk 2]
```diff
@@ -50,8 +50,13 @@ export default class List extends React.Component<ListProps, ListState> {
     return values.findIndex(a => this.getId(a) === this.getId(selectedItem));
   };
 
-  getId = (item: Object | string): string =>
-    this.props.getTextToReplace(item).text;
+  getId = (item: OutputItem | string): string => {
+    if (typeof item === 'string' || typeof item.key === 'undefined') {
```
Comment (jukben) [inline src/List.jsx]: Why do we need also check for `typeof item === 'string'`?

File: src/Textarea.jsx (hunk 1 above)
Comment (jukben) [reply src/Textarea.jsx]: Yeah! You are definitely right. I'm just wondering why not to use `Array.isArray`?

File: __tests__/index.spec.js [hunk 3]
```diff
@@ -320,3 +320,50 @@ describe('using ref to the ReactTextareaAutocomplete to call methods', () => {
     expect(actual).toBe(expected);
   });
 });
+
+describe('data provided with keys', () => {
+  const mockedChangeFn = jest.fn();
+  const mockedSelectFn = jest.fn();
+  const mockedCaretPositionChangeFn = jest.fn();
+
+  const createRtaComponent = override => (
```
Comment (jukben) [inline __tests__/index.spec.js]: I would probably remove the override as soon as we don't use it. What do you think?

File: src/List.jsx (hunk 2 above)
Comment (jukben) [reply src/List.jsx]: I see. It's because of a flow, right? Then it's fine. 💪

File: __tests__/index.spec.js (hunk 3 above)
Comment (alecklandgraf) [reply __tests__/index.spec.js]: I was trying to mock `console.error` and use `override` to pass data without keys, but I couldn't get it working. 
If you don't include keys and have a duplicate output, react throws an warning about duplicate keys to `console.error`. The test output is missing this, so we're 👍

File: src/Textarea.jsx (hunk 1 above)
Comment (alecklandgraf) [reply src/Textarea.jsx]: `Array.isArray` isn't supported on IE8, but I just checked and React 15 and up don't support IE8 anyways so I'll update to `Array.isArray`.

File: __tests__/index.spec.js [hunk 4]
```diff
@@ -320,3 +327,121 @@ describe('using ref to the ReactTextareaAutocomplete to call methods', () => {
...
+    ).toMatchSnapshot();
+  });
+
+  it('should display all items', () => {
+    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
+      3
+    );
+  });
+
+  it('should use the key property from the data obejct', () => {
```
Comment (alecklandgraf) [inline __tests__/index.spec.js]: I added an explicit test for the keys here

File: __tests__/index.spec.js [hunk 5]
```diff
@@ -78,6 +79,12 @@ describe('object-based items', () => {
     );
   });
 
+  it('should use the create a key from the output function if no key is provided', () => {
+    const items = rta.find(Item);
+    expect(items.at(0).key()).toEqual('___happy_face___');
+    expect(items.at(1).key()).toEqual('___sad_face___');
```
Comment (alecklandgraf) [inline __tests__/index.spec.js]: I added a test for the default key if `key` isn't provided with the data

File: src/types.js [hunk 6]
```diff
@@ -94,3 +94,7 @@ export type TextareaState = {
   selectionStart: number,
   component: ?React$StatelessFunctionalComponent<*>,
 };
+
+export type OutputItem = {
+  key: string,
```
Comment (alecklandgraf) [inline src/types.js]: I'm new to flow, this should be a string or number, would `?string` work for that?

**Comment by alecklandgraf** [comment]
@jukben PTAL I updated to use `Array.isArray` and added explicit tests for the keys

File: src/Textarea.jsx (hunk 1 above)
Comment (jukben) [reply src/Textarea.jsx]: Yes. Thanks! Anyway, we do transpile all the code, so Babel does it for us in every case.

File: src/types.js (hunk 6 above)
Comment (jukben) [reply src/types.js]: Yeah! `?string` is fine! 👍

**Comment by jukben** [comment]
Please be aware, we already have `outputItem` (just with small O), but in your case is probably [`textToReplaceType`](https://github.com/alecklandgraf/react-textarea-autocomplete/blob/f1956ca1a5e0d11fc80f5e71ec5d5df8ad0112d6/src/types.js#L5). Could you please refactor your code? We are almost there. Thanks a lot for your cooperation.

File: README.md [hunk 7]
```diff
@@ -118,9 +118,20 @@ export default App;
...
     [triggerChar: string]: {|
-        ?output: (item: Object | string, trigger?: string) => {| text: string, caretPosition: "start" | "end" | "next" | number |} | string,
-        dataProvider: (token: string) => Promise<Array<Object | string>> | Array<Object | string>,
-        component: ReactClass<*>,
+      output?: (
+        item: Object | string,
+        trigger?: string
+      ) =>
+        | {|
+            key?: string | number,
```
Comment (jukben) [inline README.md]: I did this change but could you please change this as well to `?string` (it's the correct way how React defines `key` so we should stick with it as well as we discussed in different thread)

File: src/List.jsx [hunk 8]
```diff
@@ -50,12 +50,12 @@ export default class List extends React.Component<ListProps, ListState> {
...
   };
 
-  getId = (item: OutputItem | string): string => {
+  getId = (item: textToReplaceType | string): string => {
     if (typeof item === 'string' || typeof item.key === 'undefined') {
       return this.props.getTextToReplace(item).text;
     }
 
-    return item.key;
+    return String(item.key);
```
Comment (alecklandgraf) [inline src/List.jsx]: I had to cast to a sting to make flow happy

**Comment by alecklandgraf** [comment]
@jukben should be all set now! Thanks for the review!

File: src/List.jsx (hunk 8 above)
Comment (jukben) [reply src/List.jsx]: The flow was unsatisfied with our check for `typeof item.key === 'undefined'` so I've fixed this more elegantly I hope (with implicit conversion). 👍

**Comment by jukben** [comment]
Thanks! Let's merge this. 🚀

**Comment by alecklandgraf** [comment]
@jukben I think you'll have to merge, let me know if you want me to rebase.

image

**Comment by jukben** [comment]
@alecklandgraf I know, I just wanted to sound epic ... and I've accidentally forgotten to click the merge button 🤣 

Published as @webscopeio/react-textarea-autocomplete 2.1.0 🚀

**Pull request merged**